### ISSUE TITLE
FHSS Transmit only works with Strings, not arrays of any type

### ISSUE THREAD

**Issue by specternecter** [jgromes/RadioLib#769]
First I would like to note that the SX1276 cannot send 256 bytes. Semtech says it's 255 and I have verified this to be the max by trying to send 255 characters and then 256. 255 characters works while 256 does not.

Onto the real problem.

Personally I hate the use of "Strings" with a passion. But I have tried and tried to send arrays with FHSS and it absolutely will not work. I send encrypted messages which are encrypted using uint8_t arrays (non-official standard as opposed to char arrays which will change how encryption functions) startTransmit will not allow you to do this because it requires char arrays not uint8_t arrays. Not really a big deal right? just typecast is as startTransmit((char *) myUint8_tArray) and all should be well. But it is not. It should hop 50 channels in my case and take about 20 seconds to send 254 bytes. But is takes about 1 second and hops only 3 channels. Even changing the array from uint8_t to char globally doesn't work. Send a String of the same data and it works as expected. Takes roughly 20 seconds and uses all 50 channels. But you can't convert an array into a String in any way that the library seems to interpret correctly, and honestly I don't understand how that could possibly be ideal anyway. Can we have support for sending arrays with FHSS?

**Comment by jgromes** [comment]
> First I would like to note that the SX1276 cannot send 256 bytes. Semtech says it's 255 and I have verified this to be the max by trying to send 255 characters and then 256. 255 characters works while 256 does not.

Is the comment at the beginning of the examples the only source of this information? The packet length is checked by the library to not exceed 255 bytes, so transmit methods should return error code if you try to transmit 256 bytes or more. So it should be just na issue in the documentation. I will fix it, but please next time post a separate issue for it, thanks!

Regarding your primary problem, I don't see any reason why Strings should behave differently to byte arrays, because at the end of the day, the String gets converted - SX127x class only has `startTransmit(uint8_t* data, size_t len, uint8_t addr)`, all other overloads are in the common layer. 

>  startTransmit will not allow you to do this because it requires char arrays not uint8_t arrays

Not true, see my previous point.

>  just typecast is as startTransmit((char *) myUint8_tArray)

No, because the length of C-strings is determined by a trailing NULL. If there is no NULL delimiting the end of your array, Lord only knows how many bytes will you attempt to send until there's a NULL somewhere in the memory. 

> But is takes about 1 second and hops only 3 channels.

Probably because there's a NULL byte somewhere at the beginning of your array.

So, as per the issue template - please post your code, debug mode output (which will show what is actually being written into the SX1276 internal FIFO prior to Tx) etc. - the template exists for a reason.

**Comment by specternecter** [comment]
It actually does allow me to send 256 bytes without an error code, but my apologies.

> Regarding your primary problem, I don't see any reason why Strings should behave differently to byte arrays, because at the end of the day, the String gets converted - SX127x class only has startTransmit(uint8_t* data, size_t len, uint8_t addr), all other overloads are in the common layer.


Yeah me either but I get 

> exit status 1
invalid conversion from 'uint8_t*' {aka 'unsigned char*'} to 'const char*' [-fpermissive]>

But as for 

> Probably because there's a NULL byte somewhere at the beginning of your array.

Well, yeah because that's how encryption works. A NULL character can be an 'A' or an 'x' or anything. It takes a value and changes is to another value betwenn 0 and 255. There's no problems with Transmit or Transmit Interrupt so is the issue that we just can't send encrypted messages with FHSS then?

> So, as per the issue template - please post your code, debug mode output (which will show what is actually being written into the SX1276 internal FIFO prior to Tx) etc. - the template exists for a reason

Okay, I'll do that. My mistake. I thought this would be an easy issue to understand.

**Comment by jgromes** [comment]
> It actually does allow me to send 256 bytes without an error code

That's interesting, because LoRa mode is limited to 255 bytes. FSK mode on the other is not limited at all and can transfer "inifinite" packets using Stream transfer. FHSS **should** be limited to 255 bytes, since it also uses LoRa - I will need to see the code.

> exit status 1
invalid conversion from 'uint8_t*' {aka 'unsigned char*'} to 'const char*' [-fpermissive]>

Not possible to comment on this until I see what caused this, so I'll comment once you post the code.

**Comment by jgromes** [comment]
> considering we know it's stopping early because there's a null character

I don't know that, it's my guess, albeit an educated one. Feel free to just post the code.

**Comment by specternecter** [comment]
> I don't know that, it's my guess, albeit an educated one. Feel free to just post the code.

Yeah and you're right because It's an example of an encrypted message, which very often does start with a NULL character. This works in all other examples. Just doesn't work in FHSS because you cannot send an array. You have to convert it to a char array or it won't even compile.

`/*
   RadioLib SX127x Transmit with Frequency Hopping Example

   This example transmits packets using SX1278 LoRa radio module.
   Each packet contains up to 256 bytes of data, in the form of:
    - Arduino String
    - null-terminated char array (C-string)
    - arbitrary binary data (byte array)

   Other modules from SX127x/RFM9x family can also be used.

   For default module settings, see the wiki page
   https://github.com/jgromes/RadioLib/wiki/Default-configuration#sx127xrfm9x---lora-modem

   For full API reference, see the GitHub Pages
   https://jgromes.github.io/RadioLib/

   SX127x supports FHSS or Frequency Hopping Spread Spectrum.
   Once a hopping period is set and a transmission is started, the radio
   will begin triggering interrupts every hop period where the radio frequency
   is changed to the next channel.
*/

#include <RadioLib.h>
#include <Adafruit_NeoPixel.h>

#define LED 21
#define BTN 20 
Adafruit_NeoPixel pixels(1, LED, NEO_GRB + NEO_KHZ800); // rIOT-LITE
SX1276 radio = new Module(3/*NSS*/, 10/*IO0*/, 7/*RST*/, 2/*IO1*/); // rIOT-LITE

// or using RadioShield
// https://github.com/jgromes/RadioShield
//SX1278 radio = RadioShield.ModuleA;

// flag to indicate that a packet was received
volatile bool transmittedFlag = false;

// flag to indicate frequency must be changed
volatile bool fhssChangeFlag = false;

// the channel frequencies can be generated randomly or hard coded
// NOTE: The frequency list MUST be the same on both sides!
//float channels[50] = {902.50, 903.00, 903.50, 904.00, 904.50, 905.00, 905.50, 906.00, 906.50, 907.00, 907.50, 908.00, 908.50, 909.00, 909.50, 910.00, 910.50, 911.00, 911.50, 912.00, 912.50, 913.00, 913.50, 914.00, 914.50, 915.00, 915.50, 916.00, 916.50, 917.00, 917.50, 918.00, 918.50, 919.00, 919.50, 920.00, 920.50, 921.00, 921.50, 922.00, 922.50, 923.00, 923.50, 924.00, 924.50, 925.00, 925.50, 926.00, 926.50, 927.00};
float channels[50] = {902.50, 904.00, 905.50, 907.00, 908.50, 910.00, 911.50, 913.00, 914.50, 916.00, 917.50, 919.00, 920.50, 922.00, 923.50, 925.00, 926.50, 903.00, 904.50, 906.00, 907.50, 909.00, 910.50, 912.00, 913.50, 915.00, 916.50, 918.00, 919.50, 921.00, 922.50, 924.00, 925.50, 927.00, 903.50, 905.00, 906.50, 908.00, 909.50, 911.00, 912.50, 914.00, 915.50, 917.00, 918.50, 920.00, 921.50, 923.00, 924.50, 926.00};
int numberOfChannels = 50;

// counter to keep track of how many frequency hops were performed
int hopsCompleted = 0;

// counter that increments with each sent packet
int packetCounter = 0;

// save transmission state between loops
int transmissionState = RADIOLIB_ERR_NONE;

// Yes we start with \0 because well, encryption can put a \0 anywhere in place of the value being encrypted
uint8_t encryptedMessage[254] = {0x00,0x01,0x02,0x03,0x04,0x05,0x06,0x07,0x08,0x09,0x0A,0x0B,0x0C,0x0D,0x0E,0x0F,0x10,0x11,0x12,0x13,0x14,0x15,0x16,0x17,0x18,0x19,0x1A,0x1B,0x1C,0x1D,0x1E,0x1F,0x20,0x21,0x22,0x23,0x24,0x25,0x26,0x27,0x28,0x29,0x2A,0x2B,0x2C,0x2D,0x2E,0x2F,0x30,0x31,0x32,0x33,0x34,0x35,0x36,0x37,0x38,0x39,0x3A,0x3B,0x3C,0x3D,0x3E,0x3F,0x40,0x41,0x42,0x43,0x44,0x45,0x46,0x47,0x48,0x49,0x4A,0x4B,0x4C,0x4D,0x4E,0x4F,0x50,0x51,0x52,0x53,0x54,0x55,0x56,0x57,0x58,0x59,0x5A,0x5B,0x5C,0x5D,0x5E,0x5F,0x60,0x61,0x62,0x63,0x64,0x65,0x66,0x67,0x68,0x69,0x6A,0x6B,0x6C,0x6D,0x6E,0x6F,0x70,0x71,0x72,0x73,0x74,0x75,0x76,0x77,0x78,0x79,0x7A,0x7B,0x7C,0x7D,0x7E,0x7F,0x80,0x81,0x82,0x83,0x84,0x85,0x86,0x87,0x88,0x89,0x8A,0x8B,0x8C,0x8D,0x8E,0x8F,0x90,0x91,0x92,0x93,0x94,0x95,0x96,0x97,0x98,0x99,0x9A,0x9B,0x9C,0x9D,0x9E,0x9F,0xA0,0xA1,0xA2,0xA3,0xA4,0xA5,0xA6,0xA7,0xA8,0xA9,0xAA,0xAB,0xAC,0xAD,0xAE,0xAF,0xB0,0xB1,0xB2,0xB3,0xB4,0xB5,0xB6,0xB7,0xB8,0xB9,0xBA,0xBB,0xBC,0xBD,0xBE,0xBF,0xC0,0xC1,0xC2,0xC3,0xC4,0xC5,0xC6,0xC7,0xC8,0xC9,0xCA,0xCB,0xCC,0xCD,0xCE,0xCF,0xD0,0xD1,0xD2,0xD3,0xD4,0xD5,0xD6,0xD7,0xD8,0xD9,0xDA,0xDB,0xDC,0xDD,0xDE,0xDF,0xE0,0xE1,0xE2,0xE3,0xE4,0xE5,0xE6,0xE7,0xE8,0xE9,0xEA,0xEB,0xEC,0xED,0xEE,0xEF,0xF0,0xF1,0xF2,0xF3,0xF4,0xF5,0xF6,0xF7,0xF8,0xF9,0xFA,0xFB,0xFC,0xFD};

// this is the packet that will be sent
String longPacket = "Let's create a really long packet to trigger \
lots of hop interrupts. A packet can be up to 256 bytes long. \
This packet is 222 bytes so using sf = 9, bw = 125, timeOnAir is \
1488ms. 1488ms / (9*4.10ms) = 40 hops. Counter: ";

// this function is called when a complete packet
// is transmitted by the module
// IMPORTANT: this function MUST be 'void' type
//            and MUST NOT have any arguments!
#if defined(ESP8266) || defined(ESP32)
  ICACHE_RAM_ATTR
#endif
void setTxFlag(void) {
  transmittedFlag = true;
}

// this function is called when FhssChangeChannel interrupt occurs
// (at the beginning of each transmission)
// IMPORTANT: this function MUST be 'void' type
//            and MUST NOT have any arguments!
#if defined(ESP8266) || defined(ESP32)
  ICACHE_RAM_ATTR
#endif
void setFHSSFlag(void) {
  fhssChangeFlag = true;
}

void setup() {
  pixels.setPixelColor(0, pixels.Color(0, 0, 0));
  pixels.show();   // Send the updated pixel colors to the hardware.
  
  //gpio_set_direction(GPIO_NUM_21, GPIO_MODE_INPUT);
  
  Serial.begin();
  delay(1000);

  // begin radio on home channel
  Serial.print("[LoRa] Initializing ... ");
  int state = radio.begin(/*freq*/ channels[0],/*bw*/ 31.25,/*sf*/ 11, /*cr*/ 5, /*sw*/ 0xDD, /*pwr*/ 20, /*pre*/ 6, /*gain*/ 0); // NEW VOLKSNET STANDARD
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true);
  }

  Serial.print("[LoRa] Enabling CRC ... ");
  state = radio.setCRC(true);
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true);
  }

  // set hop period in symbols
  // this will also enable FHSS
  Serial.print("[LoRa] Setting Hopping Period...");
  state = radio.setFHSSHoppingPeriod(6);
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true);
  }

  // set the function to call when transmission is finished
  radio.setDio0Action(setTxFlag, RISING);

  // set the function to call when we need to change frequency
  radio.setDio1Action(setFHSSFlag, RISING);

  // start transmitting the first packet
  Serial.print("[LoRa] Sending first packet ... ");
  //String packet = longPacket + packetCounter;
  //String packet = (char *)packetBytes;
  //String packet = "We're sending a packet that's 256 bytes long in order to ensure it works and that we receive this message properly. At the end of this message will be a '*'. This * will be at the 256th byte, which is the one of most concern. Some data says 255 byte max  *"; 
  //transmissionState = radio.startTransmit(packet);
  //transmissionState = radio.startTransmit((char *)encryptedMessage);
  transmissionState = radio.startTransmit(encryptedMessage);
}

void loop() {
  // check if the transmission flag is set
  if (transmittedFlag == true) {
    // reset flag
    transmittedFlag = false;

    if (transmissionState == RADIOLIB_ERR_NONE) {
      //pixels.setPixelColor(0, pixels.Color(0, 0, 0));
      //pixels.show();   // Send the updated pixel colors to the hardware.
  
      // packet was successfully sent
      Serial.println(F("transmission finished!"));

    } else {
      Serial.print(F("failed, code "));
      Serial.println(transmissionState);

    }

    // The channel is automatically reset to 0 upon completion
    Serial.print("[LoRa] Radio is on channel: ");
    Serial.println(radio.getFHSSChannel());

    // print the number of hops it took
    Serial.print("[LoRa] Hops completed: ");
    Serial.println(hopsCompleted);

    // wait a second before transmitting again
    delay(2000);

    // reset the counter
    hopsCompleted = 0;

    // return to home channel before the next transaction
    //radio.setFrequency(channels[radio.getFHSSChannel() % numberOfChannels]); // if left running for more than a couple minutes they get out of sync and you'll get nothing but CRC errors and a ridiculously high frequency error. Set it to channel 0 and problem is eliminated.
    radio.setFrequency(channels[0]);

    // wait a second before transmitting again
    //delay(2000);

    // increment the packet counter
    packetCounter++;

    Serial.println(); // line space

    // send another packet
    Serial.print("[LoRa] Sending another packet ... ");

    //pixels.setPixelColor(0, pixels.Color(255, 255, 255));
    //pixels.show();   // Send the updated pixel colors to the hardware.
  
    //String packet = longPacket + packetCounter;
    //transmissionState = radio.startTransmit((char *)encryptedMessage);
    transmissionState = radio.startTransmit(encryptedMessage);
  }

  // check if we need to do another frequency hop
  if (fhssChangeFlag == true) {
    // we do, change it now
    int state = radio.setFrequency(channels[radio.getFHSSChannel() % numberOfChannels]);
    if (state != RADIOLIB_ERR_NONE) {
      Serial.print("[LoRa] Failed to change frequency, code ");
      Serial.println(state);
    }

    // increment the counter
    hopsCompleted++;

    // clear the FHSS interrupt
    radio.clearFHSSInt();

    // we're ready to do another hop, clear the flag
    fhssChangeFlag = false;
  }
}`

**Comment by specternecter** [comment]
This results in 

> Using library arduino_30379 at version 1.10.4 in folder: C:\Users\Brad\Desktop\Dropbox\Arduino\libraries\arduino_30379 
exit status 1
invalid conversion from 'uint8_t*' {aka 'unsigned char*'} to 'const char*' [-fpermissive]

but changing "transmissionState = radio.startTransmit(encryptedMessage);" to "transmissionState = radio.startTransmit((char *)encryptedMessage);" will compile but cuts off the message due to the NULL character as we discussed.

**Comment by specternecter** [comment]
11:59:19.971 -> ESP-ROM:esp32c3-api1-20210207
11:59:19.971 -> [LoRa] Initializing ... success!
11:59:19.971 -> [LoRa] Enabling CRC ... success!
11:59:19.971 -> [LoRa] Setting Hopping Period...success!
11:59:19.971 -> [LoRa] Sending first packet ... transmission finished!
11:59:21.189 -> [LoRa] Radio is on channel: 3
11:59:21.189 -> [LoRa] Hops completed: 3
11:59:23.157 -> 
11:59:23.157 -> [LoRa] Sending another packet ... transmission finished!
11:59:24.376 -> [LoRa] Radio is on channel: 3
11:59:24.376 -> [LoRa] Hops completed: 3
11:59:26.345 -> 
11:59:26.345 -> [LoRa] Sending another packet ... transmission finished!
11:59:27.564 -> [LoRa] Radio is on channel: 3
11:59:27.564 -> [LoRa] Hops completed: 3

**Comment by specternecter** [comment]
Thats with char array

**Comment by specternecter** [comment]
now when we do transmissionState = radio.startTransmit(encryptedMessage, sizeof(encryptedMessage)); (my mistake on that, I tried many things and forgot to set it back and didn't catch it)

we get 

> 12:01:47.314 -> [LoRa] Sending another packet ... transmission finished!
12:02:07.173 -> [LoRa] Radio is on channel: 50
12:02:07.173 -> [LoRa] Hops completed: 50

Now what the hell is this? Because

`// The channel is automatically reset to 0 upon completion
    Serial.print("[LoRa] Radio is on channel: ");
    Serial.println(radio.getFHSSChannel());`

Maybe this is why I'm not receiving anything when sending a byte array over FHSS but can receive a String just fine?

**Comment by specternecter** [comment]
Yeah that seems to be it. I get nothing but a CRC error and a really high frequency error. I confused the issue with not being able to transmit uint8_t arrays because I tried many things and can't get a message, but eventually I do start getting messages. They just are infrequent or can't be read because it's for some reason going to channel 50 instead of 0.

**Comment by jgromes** [comment]
> Maybe this is why I'm not receiving anything when sending a byte array over FHSS but can receive a String just fine?

I will have to see the receiver code to comment on that.

**Comment by specternecter** [comment]
Well I mean it keeps going to channel 50 which is going to overflow to 0 since 50 channels would only go up to [49]. After about 10 minutes the transmitter stops going to channel "50" and starts going to channel 0. That's when it starts working. The receiver code always starts at channel 0 so as soon as the transmitter decides to do the same it works. I can post the receiver code but it's doing exactly what it should the whole time

**Comment by specternecter** [comment]
I think it might be a bug with the SX1276 module that just happened to appear here because I've never experienced this channel 50 issue before and I don't see anywhere in the code where it tells the sx1276 to do that. It's doing it on it's own. I unplugged power for 5 minutes and now it doesn't seem to do it anymore. Maybe a coincidence that this happened the first time i transitioned FHSS from testing with Strings to testing with arrays? Nothing else makes sense

**Comment by jgromes** [comment]
I tried to run the original example and everything is working, so I don't think there is any issue in the library.

I have to admit I did not try it with 50 channels because where I am, it's very illegal to transmit without license in the bands you are using. That's why the example only uses 5 bands in the ISM unlicensed band.

I also want to take a moment to quote the license here:

"(...) IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE."

Unless you have a license to use all of the bands you have, you are likely violating frequency spectrum regulations.
